pkg/graveler/settings/manager.go

```diff
@@ -0,0 +1,183 @@
+	"google.golang.org/protobuf/proto"
+)
+
+const (
+	settingsSuffixTemplate = "%s/settings/%s.json"
+	cacheSize              = 100_000
+	defaultCacheExpiry     = 1 * time.Second
+)
+
+type cacheKey struct {
+	RepositoryID graveler.RepositoryID
+	Key          string
+}
+
+// Manager is a key-value store for Graveler repository-level settings.
+// Each setting is stored under a key, and it can be any proto.Message.
+type Manager struct {
+	refManager                  graveler.RefManager
+	branchLock                  graveler.BranchLocker
+	blockAdapter                block.Adapter
+	committedBlockStoragePrefix string
+	cache                       cache.Cache
+}
+
+type ManagerOption func(m *Manager)
+
+func WithCacheExpiry(expiry time.Duration) ManagerOption {
+	return func(m *Manager) {
+		m.cache.SetExpiry(expiry)
+	}
+}
+
+func NewManager(refManager graveler.RefManager, branchLock graveler.BranchLocker, blockAdapter block.Adapter, committedBlockStoragePrefix string, options ...ManagerOption) *Manager {
+	m := &Manager{
+		refManager:                  refManager,
+		branchLock:                  branchLock,
+		blockAdapter:                blockAdapter,
+		committedBlockStoragePrefix: committedBlockStoragePrefix,
+		cache: cache.NewCache(cacheSize, defaultCacheExpiry, func() time.Duration {
+			return 0
+		}),
+	}
+	for _, o := range options {
+		o(m)
+	}
+	return m
+}
+
+// Save persists the given proto.Message under the given repository and key.
+func (m *Manager) Save(ctx context.Context, repositoryID graveler.RepositoryID, key string, message proto.Message) error {
+	repo, err := m.refManager.GetRepository(ctx, repositoryID)
+	if err != nil {
+		return err
+	}
+	messageBytes, err := proto.Marshal(message)
+	if err != nil {
+		return err
+	}
+	err = m.blockAdapter.Put(ctx, block.ObjectPointer{
+		StorageNamespace: string(repo.StorageNamespace),
+		Identifier:       fmt.Sprintf(settingsSuffixTemplate, m.committedBlockStoragePrefix, key),
+		IdentifierType:   block.IdentifierTypeRelative,
+	}, int64(len(messageBytes)), bytes.NewReader(messageBytes), block.PutOpts{})
+	if err != nil {
+		return err
+	}
+	logger := logging.FromContext(ctx)
+	if logger.IsTracing() {
+		logger.
+			WithField("repo", repositoryID).
+			WithField("key", key).
+			WithField("setting", protojson.Format(message)).
+			Trace("saving repository-level setting")
+	}
+	return nil
+}
+
+// Get fetches the setting under the given repository and key.
+// The result is placed in message.
+// The result is eventually consistent: it is not guaranteed to be the most up-to-date setting. The cache expiry period is 1 second.
+func (m *Manager) Get(ctx context.Context, repositoryID graveler.RepositoryID, key string, message proto.Message) error {
+	qualifiedKey, err := json.Marshal(cacheKey{
+		RepositoryID: repositoryID,
+		Key:          key,
+	})
+	if err != nil {
+		return err
+	}
+	messageBytes, err := m.cache.GetOrSet(string(qualifiedKey), func() (v interface{}, err error) {
```

Comment:
Not sure I understand.  `GetOrSet` returns a value of type `interface{}`; why do we need to serialize to put something in and deserialize when we get it out?  E.g. [auth `LRUCache.GetCredential`](https://github.com/treeverse/lakeFS/blob/56ccefd3081af88450475d0992089aa6df39bf48/pkg/auth/cache.go#`L37) doesn't serialize.

Reply:
Change to use the struct as the key